crypto/evp/pmeth_gn.c

```diff
@@ -153,10 +153,15 @@ int EVP_PKEY_gen(EVP_PKEY_CTX *ctx, EVP_PKEY **ppkey)
             evp_pkey_export_to_provider(ctx->pkey, ctx->libctx,
                                         &tmp_keymgmt, ctx->propquery);
 
-        if (keydata == NULL)
+        if (tmp_keymgmt == NULL)
             goto not_supported;
-        ret = evp_keymgmt_gen_set_template(ctx->keymgmt,
-                                           ctx->op.keymgmt.genctx, keydata);
+        /*
+         * If there is no keydata, then the given pkey is empty, so
+         * there is no template to set
+         */
+        if (keydata != NULL)
+            ret = evp_keymgmt_gen_set_template(ctx->keymgmt,
+                                               ctx->op.keymgmt.genctx, keydata);
```

Comment:
My [comment in #11371](https://github.com/openssl/openssl/pull/11371#pullrequestreview-394386991) got me wondering if it's correct to make the call conditional on a non-NULL `keydata`...
This is compatible with the `EVP_PKEY_METHOD` key generators we have, which simply ignore the template EVP_PKEY unless they are interested in its parameters, but that's arguably a bug (an omission) rather than an intended feature.  In that respect, the code in #11549 is actually broken, as RSA doesn't support this sort of template (i.e. it's  ignored in the `EVP_PKEY_METHOD` key generators).

Either way, these lines are only partially backward compatible.  If the template has been populated with actual key data, `evp_keymgmt_gen_set_template` will fail if OP_keymgmt_gen_set_template hasn't been implemented.

So the question remains, should `evp_keymgmt_gen_set_template` return 1 rather than 0 if OP_keymgmt_gen_set_template isn't implemented?  That would be fully backward compatible, but arguably wrong...

Reply:
I think you are asking a question that may not have a good answer.
Potentially breaking existing applications (even if it is crappy code) doesnt sound like a great idea.

Reply:
Ok, so basically, your answer is that `evp_keymgmt_gen_set_template` should return 1 if OP_keymgmt_gen_set_template isn't implemented, right?

However, looking at what for example `pkey_dsa_keygen` (`crypto/dsa/dsa_pmeth.c`) does, it's perfectly acceptable to pass a NULL `keydata`, so the call to `evp_keymgmt_gen_set_template` should be unconditional, subject to the backend's judgement, if OP_keymgmt_gen_set_template is implemented.